File: scripts/value_provenance.py
"""Validate nested input origins and bounded emitted dependency regions.

This checks the stated ancestry contract, never semantic equivalence or an
original source-variable claim. Missing exact producers must stay unknown.
"""

import collections


def summarize(trace):
    """Coverage only; percentages must retain unknown/omitted denominators."""
    report = trace.get('value_provenance')
    if report is None: return {}
    counts = collections.Counter(
        input_values=sum(len(f.get('value_origins', [])) for f in trace['functions']),
        output_regions=len(report['output_regions']),
        omitted_output_regions=report['omitted_output_regions'],
        compiler_loop_control_definitions=sum(d['kind'] == 'compiler_loop_control_definition'
            for f in trace['functions'] for d in f['definitions']),
        phi_transport_maps=sum(m['phase'] in ('phi_parameter_transport', 'phi_edge_transport')
            for f in trace['functions'] for m in f['local_maps']))
    for region in report['output_regions']:
        counts['storage_' + region['relation']] += 1
        node = region.get('node_ancestry', {})
        counts['node_' + node.get('relation', 'unrecorded')] += 1
        counts['node_inlined'] += bool(node.get('inlined'))
        counts['node_cloned'] += bool(node.get('cloned'))
        counts['node_multi_origin'] += len(node.get('inputs', [])) > 1
        counts['node_incomplete'] += bool(node.get('incomplete', True))
        counts['node_kind_' + region['kind']] += 1
    for binding in report['bindings']:
        for role in binding['classifications']: counts['binding_' + role] += 1
    return dict(sorted(counts.items()))


def validate(trace, source=None):
    report = trace.get('value_provenance')
    if report is None:  # Historical schema-1 artifacts remain readable.
        return []
    errors = []
    boundaries = None
    if source is not None:
        boundaries = {0}
        offset = 0
        for character in source.decode('utf-8'):
            offset += len(character.encode('utf-8'))
            boundaries.add(offset)

    def require(ok, reason):
        if not ok and len(errors) < 20:
            errors.append(reason)

    require(report['schema_version'] == 1, 'unsupported value provenance')
    final = {row['binding_id'] for row in trace['final_bindings']}
    all_sites = {}
    functions_by_id = {}
    for index, function in enumerate(trace['functions']):
        if 'function_id' in function:
            key = function['function_id']
            require(key not in functions_by_id, 'duplicate input function identity')
            functions_by_id[key] = function
        sites = {(s['block'], s['statement_index']): s for s in function['lifted_statements']}
        for (block, statement), site in sites.items():
            all_sites[f'f{index}:b{block}:s{statement}'] = site
        nodes = function.get('value_origins', [])
        require([n['id'] for n in nodes] == list(range(len(nodes))), 'value IDs not contiguous')
        paths = set()
        known = {row['id'] for row in function['registers'] + function['definitions']}
        for node in nodes:
            location = (node['block'], node['statement_index'])
            key = (*location, tuple(node['path']))
            require(key not in paths, 'duplicate value path')
            paths.add(key)
            require(len(node['path']) <= 259 and node['path'][0] in (0, 1), 'invalid value path')
            if not function['dropped_records']:
                require(location in sites, 'nested value has no instruction cluster')
                require(node['binding_id'] is None or node['binding_id'] in known, 'unknown nested SSA binding')
            for child in node['children']:
                valid = node['id'] < child < len(nodes)
                require(valid, 'cyclic or missing value child')
                if valid:
                    nested = nodes[child]
                    require(nested['path'][:-1] == node['path'] and
                            (nested['block'], nested['statement_index']) == location,
                            'child escaped input value occurrence')
        for event in function.get('inline_events', []):
            require(event['exact_final_value_mapping'] is False, 'inline promoted to exact final producer')
    require(set(report['source_sites']) == set(all_sites), 'source site inventory differs')
    for key, site in report['source_sites'].items():
        expected = all_sites.get(key)
        if expected:
            require(site['instruction_pcs'] == expected['instruction_pcs'] and
                    site['source_lines'] == expected['source_lines'], 'invented input PC or line')
    require(len(report['output_regions']) <= report['limits']['output_regions'], 'region budget exceeded')
    require(report['visited_dependencies'] <= report['limits']['work'], 'dependency work budget exceeded')
    rows = {row['binding_id']: row for row in report['bindings']}
    require(set(rows) == final, 'binding projection inventory differs')
    for row in [*rows.values(), *report['output_regions']]:
        require(len(row['source_sites']) <= report['limits']['sites_per_region'], 'site budget exceeded')
        require(set(row['source_sites']) <= all_sites.keys(), 'unknown source site')
    for region in report['output_regions']:
        require(region['exact_value_producer'] is False, 'dependency promoted to exact producer')
        require(region['start_byte'] <= region['end_byte'], 'reversed output region')
        require(set(region['bindings']) <= final, 'region references missing binding')
        expected = sorted({site for binding in region['bindings'] if binding in rows
                           for site in rows[binding]['source_sites']})[:report['limits']['sites_per_region']]
        require(region['source_sites'] == expected, 'region dependency union differs')
        require(region['relation'] == ('storage_dependency_ancestry' if expected else 'unknown'),
                'unknown region promoted to input origin')
        require(bool(expected) or region['incomplete'], 'unattributed region claimed complete')
        node = region.get('node_ancestry')
        if node is not None:
            require(node['exact_value_producer'] is False, 'node ancestry promoted to exact value identity')
            require(len(node['inputs']) <= report['node_input_limit'], 'node origin budget exceeded')
            keys = []
            for origin in node['inputs']:
                site = report['source_sites'].get(origin['source_site'])
                require(site is not None, 'node references unknown input site')
                if site is None: continue
                require(site['function_id'] == origin['function_id'], 'node escaped its input function')
                key = (origin['function_id'], site['block'], site['statement_index'], origin['value_origin'])
                require(key not in keys, 'duplicate node input origin')
                keys.append(key)
                if origin['value_origin'] is not None:
                    values = functions_by_id.get(origin['function_id'], {}).get('value_origins', [])
                    index = origin['value_origin']
                    valid = type(index) is int and 0 <= index < len(values)
                    require(valid, 'unknown retained value occurrence')
                    if valid:
                        require((values[index]['block'], values[index]['statement_index']) ==
                                (site['block'], site['statement_index']), 'retained value escaped input statement')
            expected_relation = ('retained_node_ancestry' if node['inputs'] else
                                 'synthesized_node' if node['synthesized_by'] else 'unknown')
            require(node['relation'] == expected_relation, 'unsupported node origin relation')
            require(node['synthesized_by'] in (None, 'statement_deinline', 'expression_deinline',
                                             'arithmetic_deinline', 'terminal_synthesis'), 'unknown node synthesis producer')
            require(expected_relation != 'unknown' or node['incomplete'], 'missing node origin claimed complete')
            require(type(node['cloned']) is bool and type(node['inlined']) is bool, 'invalid node history flags')
            require(not (node['cloned'] or node['inlined']) or expected_relation != 'unknown' or node['incomplete'],
                    'node history has no evidence')
        if source is not None:
            start, end = region['start_byte'], region['end_byte']
            require(0 <= start <= end <= len(source), 'output region outside source')
            require(start in boundaries and end in boundaries, 'output region splits UTF-8')
    return errors

File: luau-lifter/src/value_provenance.rs
//! Bounded projection from immutable input value graphs to exact output regions.
//! These are dependency/source ancestry relations, not value-identity proofs.
use std::collections::{BTreeMap, BTreeSet};
use ast::{RcLocal, emission_map::EmissionMap};
use cfg::provenance::FunctionTrace;
use serde_json::{Value, json};

const SITE_LIMIT: usize = 64;
const WALK_LIMIT: usize = 256;
const WORK_LIMIT: usize = 2_000_000;

pub fn report(traces: &[Box<FunctionTrace>], locals: &BTreeMap<u64, RcLocal>, map: &EmissionMap,
              producers: &[ast::local_producers::Pass]) -> Value {
    let mut sites = BTreeMap::new();
    let mut definition_sites = BTreeMap::new();
    let mut dependencies: BTreeMap<u64, BTreeSet<u64>> = BTreeMap::new();
    let mut input_parameters = BTreeSet::new();
    let mut compiler_temporaries = BTreeSet::new();
    let mut incoming_captures = BTreeSet::new();
    let mut incomplete_ids = BTreeSet::new();
    let function_indices: BTreeMap<_, _> = traces.iter().enumerate()
        .map(|(index, trace)| (trace.function_id.as_str(), index)).collect();
    let introduced: BTreeSet<_> = producers.iter().flat_map(|p| &p.ledger.records).map(|r| r.binding_id.as_str()).collect();
    for (function, trace) in traces.iter().enumerate() {
        for site in trace.lifted.values() {
            let id = format!("f{function}:b{}:s{}", site.block, site.index);
            sites.insert(id, json!({"function_id": trace.function_id, "prototype": trace.prototype,
                "block": site.block, "statement_index": site.index,
                "instruction_pcs": site.pcs, "source_lines": site.lines, "instruction_role": site.kind}));
        }
        for register in trace.registers.values() {
            if register.kind == "parameter" { input_parameters.insert(register.id); }
            if register.kind == "incoming_upvalue" { incoming_captures.insert(register.id); }
        }
        for definition in trace.definitions.values() {
            if definition.kind == "compiler_loop_control_definition" { compiler_temporaries.insert(definition.id); }
            if let Some(statement) = definition.statement {
                let site = format!("f{function}:b{}:s{statement}", definition.block);
                if sites.contains_key(&site) { definition_sites.insert(definition.id, site); }
                else { incomplete_ids.insert(definition.id); }
            }
            dependencies.entry(definition.id).or_default().extend(&definition.dependencies);
            if trace.dropped_records != 0 { incomplete_ids.insert(definition.id); }
        }
        // Mapping is ancestry, not equality: keep both contributors even when
        // register reuse made their values or source declarations different.
        for event in &trace.maps {
            dependencies.entry(event.to).or_default().insert(event.from);
        }
        for event in &trace.inlines {
            for sink in &event.consumer_bindings {
                dependencies.entry(*sink).or_default().insert(event.producer);
            }
        }
    }
    let mut fuel = WORK_LIMIT;
    let mut output_roles: BTreeMap<u64, BTreeSet<&str>> = BTreeMap::new();
    for token in &map.bindings { output_roles.entry(token.binding_id).or_default().insert(token.role); }
    for region in &map.regions {
        if region.kind == "closure" {
            for binding in &region.bindings { output_roles.entry(*binding).or_default().insert("closure_capture"); }
        }
    }
    let mut binding_sites = BTreeMap::new();
    let mut binding_rows = Vec::new();
    for (&id, local) in locals {
        let local = local.0.lock();
        let mut pending = local.3.as_ref().map(|l| l.definitions.clone()).unwrap_or_default();
        let mut seen = BTreeSet::new();
        let mut found = BTreeSet::new();
        let mut incomplete = local.3.as_ref().is_none_or(|l| l.incomplete);
        while let Some(origin) = pending.pop() {
            if seen.contains(&origin) { continue; }
            if fuel == 0 || seen.len() == WALK_LIMIT { incomplete = true; break; }
            fuel -= 1;
            seen.insert(origin);
            incomplete |= incomplete_ids.contains(&origin);
            if let Some(site) = definition_sites.get(&origin) { found.insert(site.clone()); }
            if found.len() > SITE_LIMIT { incomplete = true; found.pop_last(); }
            if let Some(edges) = dependencies.get(&origin) { pending.extend(edges.iter().rev()); }
        }
        let roles = output_roles.get(&id).cloned().unwrap_or_default();
        let mut classification = Vec::new();
        if roles.contains("parameter") { classification.push("output_parameter"); }
        if roles.contains("iteration_binding") { classification.push("output_iteration_binding"); }
        if roles.contains("closure_capture") { classification.push("closure_constructor_capture_ancestry"); }
        if local.2.iter().any(|b| matches!(b.origin, ast::BindingOrigin::DebugLocal { .. })) {
            classification.push("recorded_source_local");
        }
        if local.4.conditional_result { classification.push("inferred_conditional_result"); }
        if local.3.as_ref().is_some_and(|l| l.definitions.iter().any(|id| compiler_temporaries.contains(id))) {
            classification.push("compiler_loop_temporary_ancestry");
        }
        if local.3.as_ref().is_some_and(|l| l.definitions.iter().any(|id| incoming_captures.contains(id))) {
            classification.push("incoming_capture_ancestry");
        }
        if introduced.contains(format!("b{id}").as_str()) { classification.push("recorded_emitter_synthesis"); }
        if local.3.as_ref().is_some_and(|l| l.copied_local_metadata) {
            classification.push("copied_local_metadata_ancestry");
        }
        if seen.iter().any(|id| input_parameters.contains(id)) { classification.push("parameter_ancestry"); }
        if classification.is_empty() { classification.push("unclassified"); }
        binding_rows.push(json!({"binding_id": format!("b{id}"), "classifications": classification,
            "source_sites": found, "incomplete": incomplete,
            "storage_and_source_identity_distinct": true}));
        binding_sites.insert(id, (found, incomplete));
    }
    let regions = map.regions.iter().map(|region| {
        let mut found = BTreeSet::new();
        let mut incomplete = false;
        for binding in &region.bindings {
            if let Some((sites, partial)) = binding_sites.get(binding) {
                incomplete |= partial;
                for site in sites {
                    found.insert(site.clone());
                    if found.len() > SITE_LIMIT { incomplete = true; found.pop_last(); }
                }
            } else { incomplete = true; }
        }
        let empty = found.is_empty();
        let mut node_inputs = Vec::new();
        let mut node_incomplete = false;
        if let Some(data) = &region.origin.0 {
            node_incomplete = data.incomplete;
            for input in &data.inputs {
                let Some(&function) = function_indices.get(input.function.as_ref()) else {
                    node_incomplete = true; continue;
                };
                let site = format!("f{function}:b{}:s{}", input.block, input.statement);
                let trace = &traces[function];
                let valid_value = input.value.is_none_or(|id| trace.values.get(id).is_some_and(|v|
                    v.block == input.block && v.statement == input.statement));
                if !sites.contains_key(&site) || !valid_value {
                    node_incomplete = true; continue;
                }
                node_inputs.push(json!({"function_id": input.function.as_ref(),
                    "value_origin": input.value, "source_site": site}));
            }
        }
        let synthesized = region.origin.0.as_ref().and_then(|o| o.synthesized);
        let node_relation = if node_inputs.is_empty() {
            if synthesized.is_some() { "synthesized_node" } else { "unknown" }
        } else { "retained_node_ancestry" };
        let node_ancestry = json!({"inputs": node_inputs,
            "inlined": region.origin.0.as_ref().is_some_and(|o| o.inlined),
            "cloned": region.origin.0.as_ref().is_some_and(|o| o.cloned),
            "synthesized_by": synthesized, "relation": node_relation,
            "incomplete": node_incomplete || node_relation == "unknown",
            "exact_value_producer": false});
        json!({"kind": region.kind,
            "start_byte": region.span.start.byte_offset, "end_byte": region.span.end.byte_offset,
            "bindings": region.bindings.iter().map(|id| format!("b{id}")).collect::<Vec<_>>(),
            "source_sites": found, "incomplete": incomplete || empty,
            "relation": if empty { "unknown" } else { "storage_dependency_ancestry" },
            "node_ancestry": node_ancestry,
            "exact_value_producer": false})
    }).collect::<Vec<_>>();
    json!({"schema_version": 1, "model": "input-value-graph-and-output-regions-v1",
        "limits": {"sites_per_region": SITE_LIMIT, "binding_walk": WALK_LIMIT,
            "work": WORK_LIMIT, "output_regions": ast::emission_map::REGION_LIMIT},
        "visited_dependencies": WORK_LIMIT - fuel, "work_exhausted": fuel == 0,
        "omitted_output_regions": map.omitted_regions,
        "source_sites": sites, "bindings": binding_rows, "output_regions": regions,
        "node_input_limit": ast::node_origins::INPUT_LIMIT,
        "contract": "Input value paths are immutable occurrences before SSA copy propagation. Final spans identify exact emitted syntax. Storage dependencies and retained node tags are separate ancestry relations, never exact producers, evaluation order, lifetime or source spelling. Node clone flags record actual AST copies surviving into output, not duplicated runtime execution; shared storage ancestry alone never proves cloning. Committed inline substitutions mark the installed node. Reductions union parent and child node origins. Newly rebuilt unattributed nodes and detached scalar leaves remain unknown. Explicit reconstruction producers identify synthetic nodes without claiming original call sites. Missing debug data never proves a compiler temporary.",
        "pass_contracts": ast::node_origins::PASS_CONTRACTS.iter().map(|(pass, policy)|
            json!({"pass": pass, "policy": policy, "transfers_effect_or_lifetime_proof": false})).collect::<Vec<_>>()})
}
